File: 2021/day13/index2.js
import fs from "fs";

const data = fs.readFileSync("data-small.txt", { encoding: "utf8" });
const lines = data.split("\n");

var emptyLineIndex = lines.findIndex((line) => line == "");

const coordinateLines = lines.slice(0, emptyLineIndex);
const foldInstructions = lines.slice(emptyLineIndex + 1);
// console.log(coordinateLines, foldInstructions);

var coordinates = coordinateLines.map((coord) =>
  coord.split(",").map((c) => c)
);
const width =
  coordinates.reduce((acc, curr) => (curr[0] > acc ? curr[0] : acc), 0) + 1;
const height =
  coordinates.reduce((acc, curr) => (curr[1] > acc ? curr[1] : acc), 0) + 1;
console.log(coordinates);
console.log("width:", width, "height:", height);

function createGrid(width, height, value) {
  return new Array(height).fill(value).map(() => new Array(width).fill(value));
}

var grid = createGrid(width, height, " ");

for (var [x, y] of coordinates) {
  console.log(x, y);
  grid[y][x] = "X";
}

// console.log("startGrid");
// console.table(grid);

var instructions = foldInstructions.map((foldInstruction) => {
  var command = foldInstruction.split("=")[0];
  var value = parseInt(foldInstruction.split("=")[1]);
  return [command, value];
});

function fold(grid, [command, value]) {
  if (command == "fold along y") {
    const newWidth = grid[0].length;
    const newHeight = Math.max(value, grid.length - 1 - value);
    const newGrid = createGrid(newWidth, newHeight, " ");

    for (let x = 0; x < newWidth; x++) {
      for (let y = 0; y < newHeight; y++) {
        newGrid[y][x] =
          grid[y][x] == "X" || grid[grid.length - 1 - y][x] == "X" ? "X" : " ";
      }
    }
    return newGrid;
  } else {
    const newWidth = Math.max(value, grid[0].length - 1 - value);
    const newHeight = grid.length;
    const newGrid = createGrid(newWidth, newHeight, " ");

    for (let x = 0; x < newWidth; x++) {
      for (let y = 0; y < newHeight; y++) {
        newGrid[y][x] =
          grid[y][x] == "X" || grid[y][grid[0].length - 1 - x] == "X"
            ? "X"
            : " ";
      }
    }
    return newGrid;
  }
}

function foldGridWithInstructions(grid, instructions) {
  if (instructions.length == 0) {
    return grid;
  } else {
    var instruction = instructions[0];
    grid = fold(grid, instruction);
    return foldGridWithInstructions(grid, instructions.slice(1), --i);
  }
}

console.log(instructions);

var endGrid = foldGridWithInstructions(grid, instructions);

// console.log("endGrid:");
console.table(endGrid);

File: 2021/day13/index2.py
with open("data.txt") as file:
    dots, folds = file.read().split("\n\n")
    dots = {(int(x), int(y)) for x, y in (row.split(",") for row in dots.splitlines())}
    folds = [(axis, int(pos)) for axis, pos in (row.split(" ")[-1].split("=") for row in folds.splitlines())]

print(folds)

def fold(x, y, axis, pos):
    if axis == "x":
        if x == pos:
            return None
        return (x, y) if x < pos else (2 * pos - x, y)
    else:
        if y == pos:
            return None
        return (x, y) if y < pos else (x, 2 * pos - y)

for i in range(len(folds)):
    dots = {dot for x, y in dots if (dot:=fold(x, y, *folds[i]))}
    if i == 0:
        print(len(dots)) # part 1

width = max(x for x, y in dots) + 1
height = max(y for x, y in dots) + 1
grid = [[" " for _ in range(width)] for _ in range(height)]
for x, y in dots:
    grid[y][x] = "#"
for row in grid:
    print("".join(row)) # part 2